elixir/apps/domain/lib/domain/config/definitions.ex

```diff
@@ -559,6 +559,23 @@ defmodule Domain.Config.Definitions do
   defconfig(:docker_registry, :string, default: "ghcr.io/firezone")
   defconfig(:api_url_override, :string, default: nil)
 
+  ##############################################
+  ## Sign Up
+  ##############################################
+
+  @doc """
+  String of comma separated email domains allowed to signup regardless of signup enabled state.
+  If signups are enabled, this list is ignored.
+  """
+  defconfig(:sign_up_allowed_domains, :string,
+    changeset: fn changeset, key ->
+      changeset
+      |> Domain.Validator.validate_does_not_contain(key, " ", message: "cannot contain spaces")
+      |> Domain.Validator.validate_does_not_contain(key, ",,")
+    end,
+    default: ""
+  )
```

Comment:
If we cast the value using this feature then we can have better changeset validations without a ton of extra code
```suggestion
  defconfig(:sign_up_allowed_domains, {:array, ","},
    changeset: &Domain.Validator.validate_fqdn/2,
    default: []
  )
```

Reply:
Ah, ok, I didn't realize that was possible.  Just to confirm, if we do this, then that means we'd have to use a list in the `config|dev.exs` config file, but a string for the ENV var correct?

Reply:
@bmanifold yeah, that is correct

Reply:
Gotcha.  I had originally had this using the `dump:` option and it was doing the conversion from string to list, but I figured we wouldn't want to have 2 different types/ways for this config option, but if that's not an issue then I'll get this updated.

Reply:
I tried your suggestion and it wasn't working at first, but I finally figured out that it needed one more value in the tuple `:string`.  After that it mostly worked, however, one problem that's popping up is the following:

If I give it the following ENV vars:
```
SIGN_UP_ALLOWED_DOMAINS="firezone.dev,"
SIGN_UP_ALLOWED_DOMAINS="firezone.dev,  "
```

Then I end up with:
```
Domain.Config.get_env(:domain, :sign_up_allowed_domains) #=> ["firezone.dev", nil]
```

Looking through the `elixir/apps/domain/lib/domain/config/caster.ex` file I didn't see anyway to pass an option to trim empty values during the cast, but I may have overlooked something.  I was able to get it to remove the `nil` value by adding: `dump: &Enum.filter(&1, fn val -> not is_nil(val) end)` as a `defconfig` option, but this doesn't feel like an optimal solution.  What are your thoughts here?

Reply:
@bmanifold but changeset function is called per item, correct? So we should able to use changeset to make sure it can't be `nil`

Reply:
Good call, I was only thinking of how to remove it, not how to prevent it.  I threw `validate_required` in there and it does catch the scenario I laid out above.